.github/bin/run-clang-tidy.sh

```diff
@@ -76,96 +111,110 @@ fi
+  echo ::group::Build compilation database
+  $(dirname $0)/make-compilation-db.sh
   echo "::endgroup::"
-else
-  echo "Skipping clang-tidy run: nothing to do"
+
+  # The files might be reported with various absolute prefixes that
+  # we remove to canonicalize and report as relative paths from project
+  # root.
+  readonly EXEC_ROOT="$(bazel info execution_root)"
+  readonly CURRENT_DIR="$(pwd)"
+  readonly CANONICALIZE_SOURCE_PATH_REGEX="^\(${EXEC_ROOT}\|${CURRENT_DIR}\|\.\)/"
+
+  # Helper script gets "<hash-value> <filename>" as single argument.
+  # Run clang-tidy on <filename> and store result in ${CONTENT_DIR}/<hash>
+  cat > "${CLANG_TIDY_HELPER}" <<EOF
+read OUTPUT_FILE INPUT_FILE <<< \$1
+${CLANG_TIDY} --quiet --config="$(cat ${CLANG_CONFIG_FILE})" \
+  ${ADDITIONAL_TIDY_OPTIONS[@]@Q} "\${INPUT_FILE}" 2>/dev/null \
+  | sed -e "s@${CANONICALIZE_SOURCE_PATH_REGEX}@@g" \
+  > "${CONTENT_DIR}/\${OUTPUT_FILE}.tmp"
+mv "${CONTENT_DIR}/\${OUTPUT_FILE}".tmp "${CONTENT_DIR}/\${OUTPUT_FILE}"
+EOF
+
+  # Call the helper shell script in parallel that gets the file/hash tuple
+  # as parameter.
+  echo "Processing in ${PARALLEL_COUNT} parallel invocations of ${CLANG_TIDY}"
+  cat "${FILE_HASHES_TO_PROCESS}" | \
+    xargs -P"${PARALLEL_COUNT}" -d'\n' -n1 -- bash "${CLANG_TIDY_HELPER}"
 fi
 
+# Assemble the outputs of all files of interest into one clang tidy output
+while read -r content_hash ; do
+  cat "${CONTENT_DIR}/${content_hash}"
+done < "${HASHES_OF_INTEREST}" > "${TIDY_OUT}"
+
+ln -sf "${TIDY_OUT}" "${CURRENT_TIDY_OUT}"  # Convenience location
 
 if [ -s "${TIDY_OUT}" ]; then
-  EXIT_CODE=1
-  echo "::error::There were clang-tidy warnings. Please fix"
-  echo "You find the raw output in ${TIDY_OUT}"
-  # Assemble a sed expression that matches all files that were mentioned
-  # in the clang-tidy output.
-  # Use that to filter the file list cache to leave out files with failures,
-  # but keep the ones that had no issues.
-  # That way, only files that had issues will have to be processed in
-  # subsequent clang-tidy runs.
-
-  # Extract filenames from error messages in clang-tidy output
-  sed 's/^\([a-zA-Z0-9_/.-]\+\):[0-9].*/\1/p;d' "${TIDY_OUT}" \
-    | sort | uniq > "${CLANG_TIDY_SEEN_CACHE}".files_with_issues
-
-  # Escape regex special characters in paths, then convert to delete sed expr
-  sed 's|\([/.-]\)|\\\1|g' "${CLANG_TIDY_SEEN_CACHE}".files_with_issues \
-    | sed 's|^.*$|/\0/d|' > "${CLANG_TIDY_SEEN_CACHE}".exclude_expression
-
-  # Exclude all lines in cache representing files that need reprocessing
-  sed -f "${CLANG_TIDY_SEEN_CACHE}".exclude_expression \
-    < "${CLANG_TIDY_SEEN_CACHE}".new > "${CLANG_TIDY_SEEN_CACHE}"
+  EXIT_CODE="${EXIT_CODE_ON_WARNINGS}"
+  echo "::warning::There were clang-tidy warnings. Please fix."
+  echo "::group::clang-tidy detailed output"
+  cat "${TIDY_OUT}"
+  echo "Find this output in ${CURRENT_TIDY_OUT}"
+  echo "::endgroup::"
+
+  echo "Summary (in ${SUMMARY_OUT})"
```

Comment:
You can also use `| tee filename` to save stdout to a file.

Reply:
then I rely on the existence of `tee`. I always have macs or windows in mind, which have an abysmal set of working standard tools available.